zookeeper/source/zk_nn_client.cc

```diff
@@ -380,6 +381,77 @@ void ZkNnClient::renew_lease(RenewLeaseRequestProto &req,
   }
 }
 
+void ZkNnClient::recover_file(std::string &file_path) {
+  int error_code;
+  FileZNode znode_data;
+  read_file_znode(znode_data, file_path);
+  std::uint64_t file_length = znode_data.length;
+
+  std::vector<std::string> children;
+  if (!zk->get_children(ZookeeperBlocksPath(file_path), children, error_code)) {
+    LOG(FATAL) << "[recover_file] Failed to get children for " << file_path;
+    return;
+  }
+
+  std::uint64_t size_left = file_length;
+  std::sort(children.begin(), children.end());
+  for (auto &child : children) {
+    std::vector<std::uint8_t> block_vec;
+    std::uint64_t block_id;
+    auto child_path = util::concat_path(ZookeeperBlocksPath(file_path), child);
+    if (!zk->get(child_path, block_vec, error_code, sizeof(block_id))) {
+      LOG(ERROR) << "[recover_file] failed block retrieval";
+    }
+    block_id = *reinterpret_cast<std::uint64_t *>(block_vec.data());
+    std::string block_metadata_path = get_block_metadata_path(block_id);
+    BlockZNode block_data;
+    std::vector<std::uint8_t> block_data_vec(sizeof(block_data));
+    if (!zk->get(block_metadata_path, block_data_vec, error_code,
+                 sizeof(block_data))) {
+      LOG(ERROR) << "[recover_file] Failed getting block size for "
+      << block_metadata_path << " with error " << error_code;
+    }
+    memcpy(&block_data, &block_data_vec[0], sizeof(block_data));
+    uint64_t block_size = block_data.block_size;
+
+    if (size_left < block_size) {
```

Comment:
I don't think this case would ever be executed under our current implementation of file appends, but it's probably good to keep if future implementations would use "reserve blocks" for file mutability.